app/code/Magento/PricingStorefront/Model/PriceManagement.php

```diff
@@ -0,0 +1,238 @@
+     * @param PriceRepository $priceRepository
+     * @param PriceBookRepository $priceBookRepository
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        PriceRepository $priceRepository,
+        PriceBookRepository $priceBookRepository,
+        LoggerInterface $logger
+    ) {
+        $this->priceRepository = $priceRepository;
+        $this->priceBookRepository = $priceBookRepository;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Remove prices.
+     *
+     * @param string $priceBookId
+     * @param array $productIds
+     * @return int
+     * @throws \Throwable
+     */
+    public function unassignPrices(string $priceBookId, array $productIds): int
+    {
+        return $this->priceRepository->delete($priceBookId, $productIds);
+    }
+
+    /**
+     * Assign product price to price book.
+     *
+     * @param string $bookId
+     * @param array $price
+     * @param string|null $parentId
+     * @return array
+     */
+    public function assignPrice(string $bookId, array $price, ?string $parentId = null) : array
+    {
+        $priceSearch = [];
+        $priceUpdate = [];
+        $finalKey = PriceRepository::KEY_FINAL;
+        $regularKey = PriceRepository::KEY_REGULAR;
+        // @TODO if qty not present what default value should be???
+        $qty = $price[PriceRepository::KEY_QTY] ?? 1.0000;
+        $productId = $price[PriceRepository::KEY_PRODUCT_ID] ?? null;
+
+        foreach ($price as $key => $value) {
+            if (in_array($key, $this->getPriceType())) {
+                $priceUpdate[$key . '_' . $finalKey] = $value[$finalKey] ?? null;
+                $priceUpdate[$key . '_' . $regularKey] = $value[$regularKey] ?? null;
+            }
+        }
+
+        if (empty($productId) || empty(array_filter($priceUpdate))) {
+            return $priceSearch;
+        }
+
+        $origBookId = $bookId;
+        $found = false;
+
+        if ($parentId) {
+            while (!$found && $bookId && $origBookId) {
+                try {
+                    $priceSearch = $this->getPriceRow($bookId, $productId, $qty, $parentId);
+                } catch (\Throwable $e) {
+                    $priceSearch = [];
+                }
+
+                if (!empty($priceSearch['price']) && $origBookId === $bookId) {
+                    // if new price different as actual but same as parent price - remove actual
+                    if (!empty($priceSearch['parent']) && $this->isPricesEqual($price, $priceSearch['parent'])) {
+                        $this->unassignPrices($origBookId, [$productId]);
+                        $origBookId = null;
+                    }
+
+                    // if price for required price book found - update it
+                    $found = true;
+                    continue;
+                }
+
+                if (!empty($priceSearch)) {
+                    // compare price with existing parent price and if prices the same - skip update
```

Comment:
Here is a potential problem that can be with the next case:
- assign price $5 to PB 1
- assign price $5 to PB 2 (parent: PB1)
- assign price $10 to PB 1

Expected: $10 for PB1, $5 for PB2
Actual: $10 for PB1 and PB2

Looks like we do not have an easy way to fix it and for now we can avoid this optimisation